website/versioned_docs/version-5.x/using-react-redux/annotating-connected-components-with-flow.md

```diff
@@ -0,0 +1,178 @@
+
+To begin, you need:
+
+- Flow upgraded to v0.89+
+- [Newest Flow Typed library definition for React Redux](https://github.com/flow-typed/flow-typed/blob/master/definitions/npm/react-redux_v5.x.x/flow_v0.89.x-/react-redux_v5.x.x.js)
+
+We include three major use cases for annotating `connect` with Flow:
+
+- Connecting stateless component with `mapStateToProps`
+- Connecting components with `mapDispatchToProps` of action creators
+- Connecting components with `mapStateToProps` and `mapDispatchToProps` of action creators
+
+### Connecting stateless component with `mapStateToProps`
+
+```jsx
+type OwnProps = {|
+  passthrough: number,
+  forMapStateToProps: string,
+|};
+type Props = {
+  ...OwnProps,
+  fromStateToProps: string
+};
+const Com = (props: Props) => <div>{props.passthrough} {props.fromStateToProps}</div>
+
+type State = {a: number};
+type InputProps = {
+  forMapStateToProps: string
+};
+const mapStateToProps = (state: State, props: InputProps) => {
+  return {
+    fromStateToProps: 'str' + state.a
+  }
+};
+
+const Connected = connect<Props, OwnProps, _,_,_,_>(mapStateToProps)(Com);
+```
+
+### Connecting components with `mapDispatchToProps` of action creators
+
+```jsx
+type OwnProps = {|
+  passthrough: number,
+|};
+type Props = {
+  ...OwnProps,
+  dispatch1: (num: number) => void,
+  dispatch2: () => void
+};
+class Com extends React.Component<Props> {
+  render() {
+    return <div>{this.props.passthrough}</div>;
+  }
+}
+
+const mapDispatchToProps = {
+  dispatch1: (num: number) => {},
+  dispatch2: () => {}
+};
+const Connected = connect<Props, OwnProps, _, _, _, _>(null, mapDispatchToProps)(Com);
+e.push(Connected);
+<Connected passthrough={123} />;
+```
+
+### Connecting components with `mapStateToProps` and `mapDispatchToProps` of action creators
+
+```jsx
+type OwnProps = {|
+  passthrough: number,
+  forMapStateToProps: string
+|};
+type Props = {
+  ...OwnProps,
+  dispatch1: () => void,
+  dispatch2: () => void,
+  fromMapStateToProps: number
+};
+class Com extends React.Component<Props> {
+  render() {
+    return <div>{this.props.passthrough}</div>;
+  }
+}
+type State = {a: number}
+type MapStateToPropsProps = {forMapStateToProps: string}
+const mapStateToProps = (state: State, props: MapStateToPropsProps) => {
+  return {
+    fromMapStateToProps: state.a
+  }
+}
+const mapDispatchToProps = {
+  dispatch1: () => {},
+  dispatch2: () => {}
+};
+const Connected = connect<Props, OwnProps, _, _, _, _>(mapStateToProps, mapDispatchToProps)(Com);
+```
+
+## Annotating nested higher order components with connect
+
+If you are at the unfortunate position where your component is wrapped with nested higher order component, it is probably more difficult to annotate by providing explicit type parameters, as doing so will probably require tediously deducing prop types at each layer:
+
+```jsx
+import { compose } from "redux";
+
+type AllProps = {/** ... */};
+
+compose(
+  connect(mapState, mapDispatch),
```

Comment:
[Here's an example of what I mean](https://flow.org/try/#0PTACDMBsHsHcCh4gAQFEAeBDAtgB0gKbLTjICuAzgJYB2A5shbgE4GYAmytFALm5yWQASACJVwpHtC40AVgQDGPZC2i4KiBdBq9kAJTZLkAXmSsAjmSqsAFAHJWmJXYCUAbkQ8AnriIBJOUU+dgAFZjUKAEYTZABvAB9kcGhoAC5kGjJsACMCZmR4gF8PeHZFSExWJLIaJSptZAAJAHkAYUiAHlbtcCo6AD4beGRkbrxtAhoedIMnHgA6AEFs3mY5sdwJqY6E5Hn97ppeugAaPf2A+SUCUPD1aKL+k-gXGcMF5dX16HGaSZ4uj0+v1hsgkMBkABaaEw0Hwby+ZCXII3MIRABMMV22Uq6UyOTyBWKiDKCgqVXANTqDRarXRgKOwKGIw2W2m+neSxWPDWSlZf22u3280OxzOwuR11uGKJTxebzmXK+fJ+mwFANFwMQKBhsPBTTyBDsFGQsAAFobTURYFRIJBkHMyJg7V5TWtcMQyPktL9-vN9QAxay8M6wIgKTA0LjjZjKHgW5AIoiCeNEWhXYIqO4UM6pqNh-URqOUIip7CJ6RaHQ8shGaBe5Bo9SJnwEeGt5DNWA0Jsm0zYzAALzxWVy+SKHiTjezWPioOFXZ72eeIwlgSlvciK-O80lwV76OeE80FQoJv5-2QBHQfBo7BNsyUItVbI6vf6cUK2ohADk4MgLVYY0AP-MNkCLB0aBoaAeEwPhEwTa9NljG5wJfdVUngbBoHYMhCHmJDoFjPtkBsWlIjItp0RsC8phcV4OUVT4eW+X1tkXd93CAA)

Reply:
In general $Diff should not be needed, but it could be OK to show it as a _special case_. e.g eif there are some complex cross-component requirements for props types, it might just be easier to add some instead, although in the long run I'd still claim that spreading would make things easier.